### PR TITLE
Refactoring and base work for 3.3

### PR DESCRIPTION
**Pull request by mgreter**
Pushing my latest refactoring work to public since it is currently at a working state that solves all open  `@at-root` issues, plus a few other open issues. ~~PR is still too dirty too actually commit the work to master branch. It also introduced a minor whitespace (linefeed preserving) regressions (from issue 1208), since the fix we added for it doesn't seem to be 100% correct. I will try to tackle the linefeed preserving again at a later stage (the logic for this seems to be clear so far).~~

The following issues should pass with this WIP branch:

- Error with `&` selector interpolation in `@at-root` blocks (#1043)
- & + '' needs to be wrapped in parentheses (#1170)
- @at-root and #{&} produces incorrect output (#1210)
- Unexpected behaviour with @at-root (1236)
- Regression in parsing @at-rules (#1253)
- Dynamic division with interpolation not compatible with libsass (#948)
- Two more tests for `is_superselector` functions (#963)
- libsass-todo-test/38_expression_in_at_directive

This branch also includes quite a few performance optimizations.
It brought my sass-spec test suite runtime from 27s down to 7s!

### TODO

- [x] add attribution to boost for the hash_combine function (https://github.com/boostorg/functional/blob/1bf379ab601eb52c97908cc2cc141fd6026bd708/include/boost/functional/hash/hash.hpp#L209-L213)
- [x] rebase libsass#master
- [x] back out `Improve function call return value handling` (3f2daf3)
  - fixed on master
- [x] back out `Improve deprecation handling in unquote function` (93d3328)
  - fixed on master
- [x] back out `Fix unquote handling for non string types` (f43b881)
  - fixed on master
- [x] back out `Add code improvements for SUN OS compilation` (b523c9e) to it's own PR (#1326)
  - there's still plenty of discussion about the _right_ way to fix this /cc @saper
  - moved to its own PR (#1326)
- [x] back out `Fix memory leaks and improve 'input_source' handling` (af12e9b)
  - there's still plenty of [discussion](https://github.com/mgreter/libsass/commit/af12e9b321e37367a22e6c883b6f5cee29ec6f11#commitcomment-12060509) about the _right_ way to fix this /cc @saper
  - moved to its own PR (#1327)
- [x] back out `Implement initial parts for reference combinators` (2f07ba3)
  - we don't have specs over this yet so we don't know what our requirements are
  - moved to its own PR (#1329)
- [x] pass the [parent-selector specs](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-tests/libsass/parent-selector)
  - these are regressions that pass on the current master
- [x] pass the [issue 1303 spec](https://github.com/sass/sass-spec/tree/master/spec/libsass-todo-issues/issue_1303)

### CODE REVIEW

**Comment by drewwells** [comment]
Holy balls that's an amazing speed up! Great work!

**Comment by xzyfer** [comment]
Great work. I'll dig into this over the next couple days.

I noticed some changes to cssize I need to look at closely. Currently it's
a direct port of the ruby version and I would prefer we don't drift. (On my
phone, I haven't looked at the code yet)

**Comment by xzyfer** [comment]
Looked at the cssize changes, they're fine :+1:

**Comment by onedayitwillmake** [comment]
Really liking the changes in this branch, everything is really making a lot more sense as someone less familiar with the code base!

**Comment by xzyfer** [comment]
@mgreter if you can get this ready for review in a week I'm cool with
shipping 3.3 soon.

**Comment by mgreter** [comment]
@onedayitwillmake Thanks a lot for your feedback! I didn't actually do much work in the parser/eval before, so I had the get myself familiar with it first too. I see some more work in pseudo selectors, since they trigger specific "is_superselector" behavior. But we may postpone this after an initial release.

@xzyfer I'd say lets set a due date and I will see how much I can get in. I'm now at a stage where I think I have parent selectors covered quite well (took me about 3 days to get it right and to wrap my head around it). Unfortunately I will not have that much time over the next weekends.

**Comment by mgreter** [comment]
Made some major break thoughs by removing the re-parsing step in ruleset expander. Basically re-implemented the complete selector parsing, expanding and evaluating part. Big performance improvements in the loop and number evaluation code parts and also a lot of smaller performance improvements from parser, expand and eval refactoring.

This branch should now pass the following open issues:

https://github.com/sass/libsass/issues/941, https://github.com/sass/libsass/issues/948, https://github.com/sass/libsass/issues/1029, https://github.com/sass/libsass/issues/1043, https://github.com/sass/libsass/issues/1061, https://github.com/sass/libsass/issues/1170, https://github.com/sass/libsass/issues/1210, https://github.com/sass/libsass/issues/1214, https://github.com/sass/libsass/issues/1255, https://github.com/sass/libsass/issues/1257, https://github.com/sass/libsass/issues/1260, https://github.com/sass/libsass/issues/1269

Now also passes these `is_superselector` tests:

- has_is_superselector_of_subset_host
- host_context_is_superselector_of_subset_host
- host_is_superselector_of_subset_host
- matches_is_superselector_of_subset_matches
- nth_match_is_not_superselector_of_nth_last_match
- nth_match_is_not_superselector_of_nth_match_with_different_arg

There is one minor white-space issue, that prevents

- https://github.com/sass/libsass/issues/1214 and basic\38_expressions_in_at_directives

Example:

```css
@mixin dummy {
  @keyframes {
    @content;
  }
}
@include dummy {
  foo, bar { baz: test; }
}
```

Ruby sass output (compressed!):
```css
@keyframes{foo, bar{baz:test}}
```

Current libsass output (compressed!):
```css
@keyframes{foo,bar{baz:test}}
```

IMHO libsass is correct here to remove the space after the comma separator.
@xzyfer if you agree here, I will commit the spec changes once the PR is merged!?

**Comment by mgreter** [comment]
@xzyfer cleaned up the most obvious stuff but all in all I'm pretty much done for the first step. So this is pretty much open for review and comments now. I added a few crutches to solve a few edge cases for now. A complete solution for these edge cases will need some more work. These are namely:

- wrapper argument for `is_superselector` evaluating
- use of `remove_parent_selectors` in `is_superselector`
- passing of `is_root` with block parsing function
- few parsing functions are not coded very optimal

Other than that I feel pretty confortable with the final result!
Should be a solid base to work towards the next 3.3 release!

ToDo: need to check if parser errors out correctly for missing scope brackets and semicolons ...

File: ast.hpp [hunk 1]
```diff
@@ -53,6 +53,13 @@
 namespace Sass {
   using namespace std;
 
+  template <typename T>
```
Comment (xzyfer) [inline ast.hpp]: Please add the correct attribution to boost.
https://github.com/boostorg/functional/blob/1bf379ab601eb52c97908cc2cc141fd6026bd708/include/boost/functional/hash/hash.hpp#L209-L213

**Comment by xzyfer** [comment]
@mgreter can you look to stablize this PR this week. We should move future iterations to a new PR. This PR is already way too big.

Once it's stable I'll commit some time to digging through it. I want to start getting some early 3.3 betas out in the next two weeks. I suspect we're going to have some pain getting 3.3.

**Comment by xzyfer** [comment]
I can confirm sass-spec runs ~2x as fast.

File: ast.hpp [hunk 2]
```diff
@@ -183,6 +192,10 @@ namespace Sass {
       elements_.insert(elements_.begin(), element);
       return *this;
     }
+    size_t size()
```
Comment (xzyfer) [inline ast.hpp]: We already have `length`. Please remove this.

File: ast.hpp [hunk 3]
```diff
@@ -1786,7 +1791,21 @@ namespace Sass {
   public:
     Type_Selector(ParserState pstate, string n)
     : Simple_Selector(pstate), name_(n)
-    { }
+    {
+      size_t pos = n.find('|');
```
Comment (xzyfer) [inline ast.hpp]: There's a lot of added duplication in these nodes. Can we hoist it into a shared parent class.

File: parser.hpp [hunk 4]
```diff
@@ -242,29 +248,27 @@ namespace Sass {
...
+    void parse_charset_directive();
     Media_Block* parse_media_block();
     List* parse_media_queries();
     Media_Query* parse_media_query();
     Media_Query_Expression* parse_media_expression();
-    Feature_Block* parse_feature_block();
-    Feature_Query* parse_feature_queries();
-    Feature_Query_Condition* parse_feature_query();
-    Feature_Query_Condition* parse_feature_query_in_parens();
-    Feature_Query_Condition* parse_supports_negation();
```
Comment (xzyfer) [inline parser.hpp]: I don't like that this has been made so shallow for a couple a reasons.

- the heavy branching makes this code hard to read
- the current implementation is [modelled after the grammar](http://www.w3.org/TR/css3-conditional/#at-supports) so it's easy to spot inconsistencies.

I'd opt to backing out this change.

Comment (xzyfer) [inline parser.hpp]: IMO we should aim to mirror the CSS grammar where possible.

**Comment by xzyfer** [comment]
I've had a look through the code. It mostly looks good to me. My main points of contention are

- I still have no idea what the distinction between String_Constant and String_Quoted is. 
  - We should update the underlying data model to better reflect their use case in a new refactor.
- The changes to how we parse feature queries make the code less readable and diverge from the spec.

In general we should try to avoid having heaps a bunch of `if/else if` switches. These severely impact code readability, where as a concise function focuses context and adds semantics (via the function name).

**Comment by xzyfer** [comment]
I'd also like to see where we can cherry-pick out isolated fixes into their own PRs. Anything we can do to cut down the size of this PR will be a benefit to us as we move forward with 3.3 betas and bug fixes.

File: ast.hpp (hunk 2 above)
Comment (mgreter) [inline ast.hpp]: Good catch, will change to `length` and remove `size`

**Comment by ekskimn** [comment]
@mgreter  Hey @onedayitwillmake  and I have been troubleshooting some issues we were seeing on our side and there are a couple of bug fixes we'd like to contribute to your branch. What's the best way to do this for you?    One fix is for the selector replace where it doesn't properly handle the specificity for a replace and the other fix is one related to child selectors where it duplicates the child selector when you assign it to a variable.

**Comment by xzyfer** [comment]
@ekskimn please do not add any more to this branch. It's already so unwiedly that GitHub wont show diffs :)

I'll be slicing of parts of this PR that don't belong. We'll aim to merge to then merge this PR at which point we can continue to iterate. If you wish to continue local development you can checkout @mgreter fork and branch of this branch. Once this is merged you can cherry pick your commits into a PR.

**Comment by ekskimn** [comment]
@xzyfer Okay makes sense.  It's just that @mgreter took in @onedayitwillmake selector function code and this kind of goes along with it.  But I will fork and branch from that branch and reapply my changes, and create a PR, does that work?

**Comment by ekskimn** [comment]
@xzyfer Also I created a cmakelists.txt for C-Lion development, do you think that that should be added to the repo?

**Comment by mgreter** [comment]
Hey @ekskimn ATM I'm not sure when I have time to re-arange this PR once again. @xzyfer I know it's a bit messy, but it's also a lot of work to have it clean all the time while doing such development (I guess you know that). @ekskimn just put your fixes somewhere on github (I also have nothing against a PR into my branch). But at least make sure we can cherry pick your commit somehow ...

I guess the code is good enough now, but there are some open remarks by @xzyfer. I already did some tests and fixes for bad input. But since we have no tests for errors, it's hard to tell how good the code quality is there. Maybe I have some time today to address the remarks by @xzyfer ...

**Comment by ekskimn** [comment]
@mgreter Hey, those PR's are now open for you.  I have one for each bug fix...i figured it would be better that way

**Comment by mgreter** [comment]
I heavily rebased this PR to improve the history. Also moved a lot of the simpler fixes/commits to the top, so they should be possible to apply to master first. But I didn't check if each one actually compiles!

@xzyfer if you think it's good enough, I'd say hit merge :rocket:

**Comment by mgreter** [comment]
I heavily rebased this PR to improve the history. Also moved a lot of the simpler fixes/commits to the top, so they should be possible to apply to master first. But I didn't check if each one actually compiles!

@xzyfer if you think it's good enough, I'd say hit merge :rocket:

**Comment by xzyfer** [comment]
@mgreter I'll be pulling out some of those smaller commit and merging to master tonight. You should be able to do a rebase in the morning (12hrs after this message).

I still want to address the changes to feature query parsing. I haven't taken another look at the PR since your rebasing this morning.

**Comment by mgreter** [comment]
@xzyfer the feature query parsing has been pulled out as requested.

**Comment by xzyfer** [comment]
Great. I'll take a look soon. I'll let you know when to rebase and merge.
Give me a couple days :)

**Comment by saper** [comment]
Wow, that's huge...

**Comment by hcatlin** [comment]
Whoa, I'd totally missed this thread... EEEEEEEEEEEE!!!!!

**Comment by mgreter** [comment]
@xzyfer There is still the issue with compressed mode (I'll copy the example from above):

```css
@mixin dummy {
  @keyframes {
    @content;
  }
}
@include dummy {
  foo, bar { baz: test; }
}
```

Ruby sass output (compressed!):
```css
@keyframes{foo, bar{baz:test}}
```

Current libsass output (compressed!):
```css
@keyframes{foo,bar{baz:test}}
```

This seems directly related to https://github.com/sass/sass/issues/1296. Therefore I propose to update the sass-specs affected by this edge case, since libsass is IMO correct and should pass those tests.

**Comment by xzyfer** [comment]
@mgreter regarding the compressed mode issue. I don't believe what you're seeing there is related to sass/sass#1296. However AFAIK that CSS isn't valid so I don't believe we are under any expectations here. 

I'd be curious to see a PR for the affected specs.

**Comment by xzyfer** [comment]
#### Status update

I've done some work on separating the concerns in this PR. I've this branch down to 12 commits (half!) in https://github.com/xzyfer/libsass/pull/7.

I've update the todo list in this PR description which when done will have this PR down to those 12 commits.

There's more work to be done, but it's we're getting closer.

**Comment by mgreter** [comment]
Pulled out the commits as requested and put them into https://github.com/sass/libsass/pull/1330

**Comment by xzyfer** [comment]
It's so pretty! :heart_eyes:!

I'll start dropping some feedback on the parser refactor over the next 48hrs. So much good stuff here!

File: ast.hpp [hunk 5]
```diff
@@ -378,13 +384,12 @@ namespace Sass {
   /////////////////
   class Media_Block : public Has_Block {
     ADD_PROPERTY(List*, media_queries)
-    ADD_PROPERTY(Selector*, selector)
   public:
     Media_Block(ParserState pstate, List* mqs, Block* b)
-    : Has_Block(pstate, b), media_queries_(mqs), selector_(0)
+    : Has_Block(pstate, b), media_queries_(mqs)
```
Comment (xzyfer) [inline ast.hpp]: Good to see `selector` gone from here

File: ast_factory.hpp [hunk 6]
```diff
@@ -15,7 +15,7 @@ namespace Sass {
     Block* new_Block(string p, size_t l, size_t s = 0, bool r = false);
     Ruleset* new_Ruleset(string p, size_t l, Selector* s, Block* b);
     Propset* new_Propset(string p, size_t l, String* pf, Block* b);
-    Feature_Query* new_Feature_Query(string p, size_t l, Feature_Query* f, Block* b);
+    Supports_Query* new_Feature_Query(string p, size_t l, Supports_Query* f, Block* b);
```
Comment (xzyfer) [inline ast_factory.hpp]: I suppose `new_Feature_Query` should also renamed.

File: ast_factory.hpp [hunk 7]
```diff
@@ -67,7 +67,7 @@ namespace Sass {
     String_Constant* new_String_Constant(string p, size_t l, string val);
     String_Constant* new_String_Constant(string p, size_t l, const char* beg);
     String_Constant* new_String_Constant(string p, size_t l, const char* beg, const char* end);
-    Feature_Query_Condition* new_Feature_Query_Condition(string p, size_t l, String* f, Expression* v);
+    Supports_Condition* new_Feature_Query_Condition(string p, size_t l, String* f, Expression* v);
```
Comment (xzyfer) [inline ast_factory.hpp]: ditto on renaming `new_Feature_Query_Condition`

File: debugger.hpp [hunk 8]
```diff
@@ -3,11 +3,34 @@
 
 #include <string>
 #include <sstream>
+#include "node.hpp"
 #include "ast_fwd_decl.hpp"
 
 using namespace std;
 using namespace Sass;
 
+/*
```
Comment (xzyfer) [inline debugger.hpp]: Remove this?

File: emitter.cpp [hunk 9]
```diff
@@ -171,7 +171,7 @@ namespace Sass {
 
   void Emitter::append_comma_separator()
   {
-    scheduled_space = 0;
+    // scheduled_space = 0;
```
Comment (xzyfer) [inline emitter.cpp]: remove this?

File: eval.cpp [hunk 10]
```diff
@@ -467,11 +489,22 @@ namespace Sass {
     }
     else
     {
-      rhs->is_delayed(false);
-      rhs = rhs->perform(this);
+      // rhs->set_delayed(false);
```
Comment (xzyfer) [inline eval.cpp]: remove else block

File: extend.cpp [hunk 11]
```diff
@@ -1530,25 +1532,18 @@ namespace Sass {
         }
       }
 
-
-
-//      DEBUG_EXEC(EXTEND_COMPOUND, printCompoundSelector(pSels, "SELS: "))
-
+      DEBUG_EXEC(EXTEND_COMPOUND, printCompoundSelector(pSels, "SELS: "))
```
Comment (xzyfer) [inline extend.cpp]: Is there value in leaving this uncommented now?

**Comment by xzyfer** [comment]
@mgreter I got overly exciting and reviewed this now. All in all it's huge improvement in readability!

I'm going to run this branch over my code companies code base. If all is well I say we :ship: !!

**Comment by xzyfer** [comment]
Heads up I've run into some segfaults and broken output when dealing with `&`. I'll work on some test cases. /cc @mgreter

**Comment by xzyfer** [comment]
@mgreter as per out chat, I've found some regressions in this PR. I'll create some sass-specs and list them here. 

This is the first regression I've found

```scss
foo > bar {
    &:baz {
        baz: true;
    }
}
```
```scss
foo bar:baz {
  baz: true; }
```

The `>` is getting dropped

**Comment by xzyfer** [comment]
```scss
foo {
    &:bar > baz {
        bam: true;
    }
}
```
```
foo:bar &:bar > baz {
  bam: true; }
```

**Comment by xzyfer** [comment]
It turns out we have very little spec coverage for `&` usage. I've added some spec coverage for the popular use cases in https://github.com/sass/sass-spec/pull/435. 

There are current 12 failing specs that pass on the current master.

**Comment by mgreter** [comment]
I've added a fix which should pass both specs you've linked!
Thanks for creating them, pretty usefull to have this covered!

File: debugger.hpp (hunk 8 above)
Comment (mgreter) [inline debugger.hpp]: Would like to keep this since it is working, but I think I got some problem within a certain source file, so I commented it out (since it is for debugging I vote to leave it in at least for now).

File: extend.cpp (hunk 11 above)
Comment (mgreter) [inline extend.cpp]: I guess it doesn't really matter. Only has any effect if DEBUG is on, and I needed to debug that information at some point, so I started to uncomment them (there are already other such statements that are uncommented). IMO it can stay that way (more debug info enabled)

File: eval.cpp (hunk 10 above)
Comment (mgreter) [inline eval.cpp]: Would like to keep it for some time, since it could be needed (and has no performance impacts).

File: emitter.cpp (hunk 9 above)
Comment (mgreter) [inline emitter.cpp]: Would like to keep it for some time, since it could be needed !?

File: ast_factory.hpp (hunk 6 above)
Comment (mgreter) [inline ast_factory.hpp]: Will do!

File: ast_factory.hpp (hunk 7 above)
Comment (mgreter) [inline ast_factory.hpp]: Will do!

File: debugger.hpp (hunk 8 above)
Comment (xzyfer) [inline debugger.hpp]: :+1:

File: emitter.cpp (hunk 9 above)
Comment (xzyfer) [inline emitter.cpp]: :+1:

File: eval.cpp (hunk 10 above)
Comment (xzyfer) [inline eval.cpp]: :+1:

File: extend.cpp (hunk 11 above)
Comment (xzyfer) [inline extend.cpp]: :+1:

**Comment by xzyfer** [comment]
I can confirm the final two todo items fixed in your latest commit. 

Can you please rebase `case Complex_Selector::REFERENCE:   cerr << "{@} "; break;` out of 9f5ef6da0e8ec9287cd800bf98174f049591b286. I ran into it when doing a bisect today. I see you removed it in 3af051e anyway but lets remove it from the source.

**Comment by xzyfer** [comment]
It's midnight here. I'll rerun this branch over our code base in the morning. If there are no other regressions I'll ship it!

**Comment by xzyfer** [comment]
I've done some quick testing and I'm see regressions in newlines in selectors. A quick example

```%foo {
  bar > bam &:before,
  bar > bam &:after {
    bar: baz;
  }
}

test {
  @extend %foo;
}
```
Result
```
bar > bam test:before,

bar > bam test:after {
  bar: baz; }
```
master
```
bar > bam test:before,
bar > bam test:after {
  bar: baz; }
```
Ruby Sass
```
bar > bam test:before, bar > bam test:after {
  bar: baz; }
```

Removing the `&` fixes the output.

```
%foo {
  bar > bam:before,
  bar > bam:after {
    bar: baz;
  }
}

test {
  @extend %foo;
}
```
Result
```
test bar > bam:before,
test bar > bam:after {
  bar: baz; }
```
master
```
test bar > bam:before,
test bar > bam:after {
  bar: baz; }
```
Ruby Sass
```
test bar > bam:before, test bar > bam:after {
  bar: baz; }
```

**Comment by xzyfer** [comment]
Also seeing new lines in selectors appearing after combinators

```
foo {
  > bar {
    > baz,
    > bam {
      bar: true;
    }
  }
}
```
Result
```
foo > bar > baz, foo >
bar > bam {
  bar: true; }
```
Master & Ruby Sass
```
foo > bar > baz,
foo > bar > bam {
  bar: true; }
```

File: inspect.cpp [hunk 12]
```diff
@@ -899,6 +900,14 @@ namespace Sass {
     Compound_Selector*           head = c->head();
     Complex_Selector*            tail = c->tail();
     Complex_Selector::Combinator comb = c->combinator();
+
+    if (c->has_line_feed()) {
```
Comment (xzyfer) [inline inspect.cpp]: This block appears to be the cause of the broken selector output reported in https://github.com/sass/libsass/pull/1249#issuecomment-120948335 and https://github.com/sass/libsass/pull/1249#issuecomment-120950668

Comment (mgreter) [inline inspect.cpp]: I already got a fix for the first issue ... they don't seem related much ...

**Comment by mgreter** [comment]
Another regression:
```css
%foo1 {
  > bar > bam &:before,
  > bar > bam &:after {
    bar: baz;
  }
}

test1 {
  @extend %foo1;
}
```

Result before the fix:
```css
> bar > bam &:before, > bar > bam &:after {
  bar: baz; }
```

Result after additional fix:
```css
> bar > bam test1:before,
> bar > bam test1:after {
  bar: baz; }
```

IMO the additional line-feed makes sense and improves readability (and is more what I would expect from the given input). So I vote to leave it that way and call it a ruby sass bug!?

**Comment by xzyfer** [comment]
> So I vote to leave it that way and call it a ruby sass bug

:+1: I don't care much for matching Sass' whitespace semantics perfectly because they're too unpredictable. Like you say as long as the output is readable it's fine.

**Comment by xzyfer** [comment]
If CI goes green :ship: it

:tada: :fireworks: :pizza:

**Comment by mgreter** [comment]
Waiting for AppVeyor then ... :rocket: :guitar:

**Comment by am11** [comment]
Congrats :congratulations:
